### ISSUE TITLE
Implement Edge Detect node

### ISSUE THREAD

**Issue by Correct-Syntax** [GimelStudio/GimelStudio#145]
Implement a core node, the Edge Detect node using either glsl or numpy/opencv. It should take in an image and output an image.

It may be a good idea to implement this as a glsl shader node for speed.

Category: FILTER

**Comment by yonMaor** [comment]
I think I may be able to help with this one.
Any other relevant details I should know?

**Comment by iwoithe** [comment]
> I think I may be able to help with this one.

Thank you. This is a big help to @Correct-Syntax and I. :slightly_smiling_face:

> Any other relevant details I should know?
- As @Correct-Syntax said, try implementing it in GLSL
- Take a look at the existing nodes for some examples of the API (found in the ``src/nodes/corenodes`` directory)
- A good example of a GLSL node is the [opacity node](https://github.com/GimelStudio/GimelStudio/tree/master/src/nodes/corenodes/filter/opacity_node)
- To actually add the edge detecting code, take a look at [this](https://learnopencv.com/edge-detection-using-opencv/) article which provides a few possible methods

Hopefully this is enough to get you started. If you have any other questions feel free to ask them here (or alternatively in the [Discord](https://discord.gg/RqwbDrVDpK) server).

**Comment by yonMaor** [comment]
So it's been a while, but I am working on this again. I've already written up some of the code.
I'm having trouble installing wxpython, are you guys aware of any issues there?

**Comment by iwoithe** [comment]
I do recall issues with installing wxPython, but it was a while ago.

To install the dependencies, run `pip install -r requirements.txt` (`python3 -m pip install -r requirements.txt` if on Linux) in the root directory of Gimel Studio. Let me know if you've tried that or it still doesn't work.

---

@Correct-Syntax it looks like this repo uses 4.2.0 while `gswidgetkit` and `gsnodegraph` use 4.1.1. Not sure if that needs to be updated

**Comment by yonMaor** [comment]
This is what I tried (pip installing the requirements), but so far no luck. I've run into several different issues, so I haven't been able to pinpoint the underlying problem.
 
I also noticed the dependency difference. 
Fixing it locally (setting both 4.1.1) did not help.

**Comment by Correct-Syntax** [comment]
@yonMaor what is your python version and platform? 

On Windows, you should be able to install wxpython for up to python 3.10, but python 3.11+ isn't going to work quite yet.

Maybe try just installing wxpython and see what comes of it:
![image](https://user-images.githubusercontent.com/60711001/213895949-64e0251e-f426-4423-b612-657024ac7eb3.png)

On Linux, you will probably have to download and install the wheel yourself from https://extras.wxpython.org/wxPython4/extras/linux/gtk3/

**Comment by Correct-Syntax** [comment]
@iwoithe the difference in version shouldn't matter. Though that could definitely be updated.

**Comment by yonMaor** [comment]
@Correct-Syntax 
On the Windows 10 PC I'm running Python 3.10 
On the Linux (Ubuntu 20.04) I'm running Python 3.9 

I'll try installing wxpython separately and let you know how it goes

**Comment by yonMaor** [comment]
No luck so far, 
I have noticed that I get conflicting dependency errors (due to the different wxpython versions)

**Comment by Correct-Syntax** [comment]
@yonMaor ok. I'll see if I can release new versions of gswidgetkit and gsnodegraph with the wxpython dependency updated.

**Comment by Correct-Syntax** [comment]
I'll mention that I just tried running the source with python 3.10 and it is pretty broken. Mainly because python/wxpython no longer silently converts floats to integers. I began updating the code to fix this, but I won't probably have the time to finish this week. 

With Python 3.9, it does work for me on Windows though.

**Comment by yonMaor** [comment]
I tried 3.9 on Ubuntu without success, but the issues were a little different. 
I'll try with a Windows 10 machine and Python 3.9.

On a different note, do you have any suggestions for an environment to test GLSL code?

**Comment by Correct-Syntax** [comment]
> On a different note, do you have any suggestions for an environment to test GLSL code?

Ideally, Gimel Studio can be used to test the GLSL (at least, that is what I did when I last worked on a GLSL node). You'll want to set this line https://github.com/GimelStudio/GimelStudio/blob/b8aa5e910434f1c91783880104fc11d2c6d9e4e7/src/gimelstudio/core/node/node.py#L38 to `` False`` so that the glsl file isn't using the cached glsl when you make a change to it, though.

**Comment by Correct-Syntax** [comment]
Everything should now be working with Python 3.10. I tested on Windows so that may be your best option.

**Comment by yonMaor** [comment]
Installing the requirements.txt directly didn't work for some reason (issues with opencv as far as I can tell).
When I tried installing the libraries manually I managed to get it to work by changing opencv to its most recent version (this could be an issue with my system though).

About testing GLSL code, using Gimel Studios works partially, since it doesn't allow me to actually debug, only to test the output. That's why I was wondering whether we have any alternative.

**Comment by yonMaor** [comment]
Also, just to make sure, I added the edge detection node under filter nodes. Can I leave it there, or is there some other location you would prefer?

**Comment by Correct-Syntax** [comment]
> issues with opencv as far as I can tell

I figured that might be the case. I believe there was a reason for locking to that specific version of opencv. It may have been because of a bug that I had encountered which may be fixed now.

> That's why I was wondering whether we have any alternative.

I guess I haven't really looked for one in the past, but https://shadered.org/ may work. That said, that is the motivation behind implementing a developer console to make debugging smoother (https://github.com/GimelStudio/GimelStudio/issues/28)
